File: ui/src/lib/indicators.ts
/**
 * 技术指标前端计算,算法与旧版 dashboard/index.html 保持一致:
 * - MA:简单移动平均
 * - MACD(12,26,9):DIF = EMA12 − EMA26,DEA = EMA9(DIF),BAR = 2×(DIF−DEA)
 * - RSI:Wilder 平滑
 * - KDJ(9,3,3)
 * - BOLL(20,2)
 */
import type { Bar } from "./api";

export type Series = (number | null)[];

export function calcMA(bars: Bar[], period: number): Series {
  const result: Series = new Array(bars.length).fill(null);
  for (let i = period - 1; i < bars.length; i++) {
    let sum = 0;
    for (let j = 0; j < period; j++) sum += bars[i - j].close;
    result[i] = sum / period;
  }
  return result;
}

function calcEMA(data: number[], period: number): Series {
  const result: Series = new Array(data.length).fill(null);
  if (data.length < period) return result;
  const k = 2 / (period + 1);
  let ema = data.slice(0, period).reduce((a, b) => a + b, 0) / period;
  result[period - 1] = ema;
  for (let i = period; i < data.length; i++) {
    ema = data[i] * k + ema * (1 - k);
    result[i] = ema;
  }
  return result;
}

export interface MacdResult {
  dif: Series;
  dea: Series;
  macd: Series;
}

export function calcMACD(closes: number[]): MacdResult {
  const fast = 12;
  const slow = 26;
  const signal = 9;
  const emaFast = calcEMA(closes, fast);
  const emaSlow = calcEMA(closes, slow);
  const dif: Series = closes.map((_, i) =>
    emaFast[i] == null || emaSlow[i] == null ? null : emaFast[i]! - emaSlow[i]!,
  );
  const difValid = dif.map((v) => v ?? 0);
  const dea = calcEMA(difValid, signal);
  const macd: Series = dif.map((d, i) =>
    d == null || dea[i] == null ? null : (d - dea[i]!) * 2,
  );
  return { dif, dea, macd };
}

export function calcRSI(closes: number[], period: number): Series {
  const result: Series = new Array(closes.length).fill(null);
  if (closes.length < period + 1) return result;
  let avgGain = 0;
  let avgLoss = 0;
  for (let i = 1; i <= period; i++) {
    const change = closes[i] - closes[i - 1];
    if (change >= 0) avgGain += change;
    else avgLoss -= change;
  }
  avgGain /= period;
  avgLoss /= period;
  result[period] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
  for (let i = period + 1; i < closes.length; i++) {
    const change = closes[i] - closes[i - 1];
    const gain = change >= 0 ? change : 0;
    const loss = change < 0 ? -change : 0;
    avgGain = (avgGain * (period - 1) + gain) / period;
    avgLoss = (avgLoss * (period - 1) + loss) / period;
    result[i] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
  }
  return result;
}

export interface KdjResult {
  k: Series;
  d: Series;
  j: Series;
}

export function calcKDJ(highs: number[], lows: number[], closes: number[], period = 9): KdjResult {
  const k: Series = new Array(closes.length).fill(null);
  const d: Series = new Array(closes.length).fill(null);
  const j: Series = new Array(closes.length).fill(null);
  let prevK = 50;
  let prevD = 50;
  for (let i = period - 1; i < closes.length; i++) {
    let hh = -Infinity;
    let ll = Infinity;
    for (let m = 0; m < period; m++) {
      hh = Math.max(hh, highs[i - m]);
      ll = Math.min(ll, lows[i - m]);
    }
    const rsv = hh === ll ? 50 : ((closes[i] - ll) / (hh - ll)) * 100;
    const curK = (2 / 3) * prevK + (1 / 3) * rsv;
    const curD = (2 / 3) * prevD + (1 / 3) * curK;
    k[i] = curK;
    d[i] = curD;
    j[i] = 3 * curK - 2 * curD;
    prevK = curK;
    prevD = curD;
  }
  return { k, d, j };
}

export interface BollResult {
  mid: Series;
  upper: Series;
  lower: Series;
}

export function calcBOLL(closes: number[], period = 20, mult = 2): BollResult {
  const mid: Series = new Array(closes.length).fill(null);
  const upper: Series = new Array(closes.length).fill(null);
  const lower: Series = new Array(closes.length).fill(null);
  for (let i = period - 1; i < closes.length; i++) {
    const slice = closes.slice(i - period + 1, i + 1);
    const ma = slice.reduce((a, b) => a + b, 0) / period;
    const variance = slice.reduce((a, b) => a + (b - ma) ** 2, 0) / period;
    const std = Math.sqrt(variance);
    mid[i] = ma;
    upper[i] = ma + mult * std;
    lower[i] = ma - mult * std;
  }
  return { mid, upper, lower };
}

File: crates/agent/src/indicators.rs
//! Indicator series missing from `astock-technical` (which ships MA/MACD
//! only): RSI (Wilder), KDJ (Chinese 9,3,3 SMA recursion) and BOLL (20,2).
//!
//! All functions are pure and deterministic; the LLM never computes numbers.

/// An indicator series aligned with the input bars.
pub type Series = Vec<Option<f64>>;

/// Three aligned indicator series (e.g. K/D/J or mid/upper/lower).
pub type TripleSeries = (Series, Series, Series);

/// Wilder RSI series. First `period` entries are `None`.
pub fn rsi_series(closes: &[f64], period: usize) -> Vec<Option<f64>> {
    let n = closes.len();
    let mut out = vec![None; n];
    if n <= period || period == 0 {
        return out;
    }
    let mut gain = 0.0;
    let mut loss = 0.0;
    for i in 1..=period {
        let diff = closes[i] - closes[i - 1];
        if diff > 0.0 {
            gain += diff;
        } else {
            loss -= diff;
        }
    }
    let mut avg_gain = gain / period as f64;
    let mut avg_loss = loss / period as f64;
    out[period] = Some(rsi(avg_gain, avg_loss));
    for i in (period + 1)..n {
        let diff = closes[i] - closes[i - 1];
        let (g, l) = if diff > 0.0 { (diff, 0.0) } else { (0.0, -diff) };
        avg_gain = (avg_gain * (period as f64 - 1.0) + g) / period as f64;
        avg_loss = (avg_loss * (period as f64 - 1.0) + l) / period as f64;
        out[i] = Some(rsi(avg_gain, avg_loss));
    }
    out
}

fn rsi(avg_gain: f64, avg_loss: f64) -> f64 {
    if avg_loss == 0.0 {
        return 100.0;
    }
    100.0 - 100.0 / (1.0 + avg_gain / avg_loss)
}

/// Chinese KDJ (9,3,3): RSV over a 9-bar window, K/D seeded at 50 with the
/// 1/3 SMA recursion, J = 3K − 2D. Returns `(K, D, J)` series aligned with
/// the input; entries before the first full window are `None`.
pub fn kdj_series(
    highs: &[f64],
    lows: &[f64],
    closes: &[f64],
    period: usize,
) -> TripleSeries {
    let n = closes.len();
    let (mut ks, mut ds, mut js): TripleSeries = (vec![None; n], vec![None; n], vec![None; n]);
    if n < period || period == 0 {
        return (ks, ds, js);
    }
    let mut k = 50.0;
    let mut d = 50.0;
    for i in (period - 1)..n {
        let lo = lows[i + 1 - period..=i]
            .iter()
            .fold(f64::INFINITY, |a, &b| a.min(b));
        let hi = highs[i + 1 - period..=i]
            .iter()
            .fold(f64::NEG_INFINITY, |a, &b| a.max(b));
        let rsv = if hi > lo {
            (closes[i] - lo) / (hi - lo) * 100.0
        } else {
            50.0
        };
        k = 2.0 / 3.0 * k + 1.0 / 3.0 * rsv;
        d = 2.0 / 3.0 * d + 1.0 / 3.0 * k;
        let j = 3.0 * k - 2.0 * d;
        ks[i] = Some(k);
        ds[i] = Some(d);
        js[i] = Some(j);
    }
    (ks, ds, js)
}

/// Bollinger bands `(mid, upper, lower)`: SMA(period) ± mult × population
/// standard deviation. Entries before the first full window are `None`.
pub fn bollinger_series(
    closes: &[f64],
    period: usize,
    mult: f64,
) -> TripleSeries {
    let n = closes.len();
    let (mut mid, mut up, mut lo): TripleSeries = (vec![None; n], vec![None; n], vec![None; n]);
    if n < period || period == 0 {
        return (mid, up, lo);
    }
    for i in (period - 1)..n {
        let window = &closes[i + 1 - period..=i];
        let mean = window.iter().sum::<f64>() / period as f64;
        let var = window.iter().map(|c| (c - mean) * (c - mean)).sum::<f64>() / period as f64;
        let sd = var.sqrt();
        mid[i] = Some(mean);
        up[i] = Some(mean + mult * sd);
        lo[i] = Some(mean - mult * sd);
    }
    (mid, up, lo)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rsi_all_gains_is_100() {
        let closes: Vec<f64> = (0..20).map(|i| 10.0 + i as f64).collect();
        let rsi = rsi_series(&closes, 14);
        assert!(rsi[..14].iter().all(Option::is_none));
        assert_eq!(rsi[14], Some(100.0));
        assert_eq!(rsi[19], Some(100.0));
    }

    #[test]
    fn rsi_known_value() {
        // Classic Wilder example shape: symmetric up/down of equal size
        // over the seed window gives RSI 50.
        let mut closes = vec![100.0];
        for i in 1..=14 {
            closes.push(if i % 2 == 0 { 100.0 } else { 101.0 });
        }
        let rsi = rsi_series(&closes, 14);
        assert!((rsi[14].unwrap() - 50.0).abs() < 1e-9);
    }

    #[test]
    fn kdj_bounds_and_seed() {
        let highs = vec![10.0; 20];
        let lows = vec![9.0; 20];
        let closes = vec![9.5; 20];
        let (k, d, j) = kdj_series(&highs, &lows, &closes, 9);
        assert!(k[..8].iter().all(Option::is_none));
        // RSV is constantly 50 and K/D seed at 50 → all stay at 50.
        assert!((k[19].unwrap() - 50.0).abs() < 1e-9);
        assert!((d[19].unwrap() - 50.0).abs() < 1e-9);
        assert!((j[19].unwrap() - 50.0).abs() < 1e-9);
    }

    #[test]
    fn bollinger_flat_series_collapses() {
        let closes = vec![7.5; 30];
        let (mid, up, lo) = bollinger_series(&closes, 20, 2.0);
        assert!(mid[..19].iter().all(Option::is_none));
        assert_eq!(mid[29], Some(7.5));
        assert_eq!(up[29], Some(7.5));
        assert_eq!(lo[29], Some(7.5));
    }
}
